jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java

```diff
@@ -311,4 +334,45 @@ private void evict(boolean direct, long excess)
         }
         return oldestEntry;
     }
+
+    private static class Bucket extends Pool<RetainableByteBuffer>
+    {
+        private final int _capacity;
+
+        Bucket(int capacity, int size)
+        {
+            super(Pool.StrategyType.THREAD_ID, size, true);
+            _capacity = capacity;
+        }
+    }
+
+    /**
+     * A variant of the {@link ArrayRetainableByteBufferPool} that
+     * uses buckets of buffers that increase in size by a power of
+     * 2 (eg 1k, 2k, 4k, 8k, etc.).
+     */
+    public static class LogBuckets extends ArrayRetainableByteBufferPool
+    {
+        public LogBuckets()
+        {
+            this(0, -1, Integer.MAX_VALUE);
+        }
+
+        public LogBuckets(int minCapacity, int maxCapacity, int maxBucketSize)
+        {
+            this(minCapacity, maxCapacity, maxBucketSize, -1L, -1L);
+        }
+
+        public LogBuckets(int minCapacity, int maxCapacity, int maxBucketSize, long maxHeapMemory, long maxDirectMemory)
+        {
+            super(minCapacity,
+                -1,
+                maxCapacity,
+                maxBucketSize,
+                maxHeapMemory,
+                maxDirectMemory,
+                TypeUtil::log2NextPowerOf2,
```

Comment:
I would not change the existing buckets-sizing algorithm without measurements. Heck, I'm not even convinced making it configurable (even only in code via an internal API) is a good idea.

We currently have little visibility over how buckets are sized in the wild so I'm worried that changing this may lead to adverse effects.

Reply:
We need to make changes like this PR so we can see how buckets are sized in the wild.     I'm definitely not proposing that `LogBuckets` be used until we see the data, but making the pool Dumpable, sampleable and extensible are all worthwhile things.

Reply:
I'm not against the power of 2 idea, but perhaps it should be stripped out to maintain, as you say, dumpability and statistics.